### PR TITLE
Implement a database backed State.

### PR DESCRIPTION
**Pull request by ultratwo**
This pull requests implements an optimized state backed by an LMDB database.

There are currently a couple of major limitations:
* The code relies on the python gc to close databases. This leaks file descriptors under pypy, causing the test suite to fail.
* Various performance optimisations are needed.
* Misuse of the API will cause silent corruption in places (e.g. calling `state_root` inside a non-db transaction) where it should throw an error.

I'm planning to remove the original optimized state soon rather than continuing to maintain it.

#### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

**Review by SamWilsn — COMMENTED** [review]
Yes, this certainly looks like code :+1:

File: src/ethereum_optimized/state_db.py [hunk 1]
```diff
@@ -0,0 +1,703 @@
...
+    def __eq__(self, other: object) -> bool:
+        """
+        Test for equality by comparing state roots.
+        """
+        if not isinstance(other, State):
+            return NotImplemented
+        return state_root(self) == state_root(other)
+
+
+def begin_db_transaction(state: State) -> None:
```
[the remark this answers is not in the document]
Comment (ultratwo) [reply src/ethereum_optimized/state_db.py:82]: We support nested transactions, both for db and non-db transactions. We don't (and for db transactions, can't) support an equivalent of `RELEASE SAVEPOINT name`, which would be very useful.

I don't think our interface corresponds to savepoints (e.g. you have to commit a child transaction before you can commit it's parent).

File: src/ethereum_optimized/__init__.py [hunk 2]
```diff
@@ -40,3 +56,9 @@ def monkey_patch_optimized_state() -> None:
     """
     for (name, value) in optimized_state_patches.items():
         setattr(normal_state, name, value)
+
+
+def monkey_patch_optimized_state_db() -> None:
+    """F"""
```
[the remark this answers is not in the document]
Comment (ultratwo) [reply src/ethereum_optimized/__init__.py]: Fixed

File: src/ethereum_optimized/state_db.py [hunk 3]
```diff
@@ -0,0 +1,703 @@
+"""F"""
```
[the remark this answers is not in the document]
Comment (ultratwo) [reply src/ethereum_optimized/state_db.py]: Fixed

**Review by SamWilsn — APPROVED** [review]

**Pull request merged**